### DISCUSSION TITLE
Brilliant Library

### DISCUSSION THREAD

**Discussion by RDunmow** [HomeSpan/HomeSpan#2] [General]
I am working on some automated blinds and wanted a library for integrating into HomeKit. I like what you have created as this fits neatly into what I have been working on.

I had been using MQTT with a bridge to integrate with Homekit, which works OK, but I much prefer this library being more direct. 

Still reading the docs (which again are great), but I would like to know if you think it's possible to continue to use the MQTT libraries alongside this for sending config updates and capturing log information, or could I do that with this library as well?

**Comment by HomeSpan** [comment]
Thanks for the comments!  From what I know about MQTT, I think it could be directly incorporated into HomeSpan as an extra module, though an external library would likely need some modifications to avoid potential interference with the HomeKit communications channel.  HomeSpan itself was meant to be an end-point device connected directly to a physical product (or an RF transmitter), but there is nothing preventing its use as an MQTT "gateway" as you suggest. 

May I ask about your current setup with the automated blinds - are you using a commercial system (Somfy?  Hunter Douglas?) or something home-built?  The reason I ask is that I'm currently implementing what I hope will be a universal Somfy controller as standalone HomeSpan device.

**Reply by HerbOtto** [reply]
I too find this library well designed - nice job!  

I’m using it with MQTT and so far no issues (PubSubClient.h).  I did have to delay MQTT initialization approx 5-10 seconds after HomeSpan's startup in order to have a consistent MQTT connection during the startup sequence.  I’m using HomeSpan with 6 Somfy shades and 1 awning using a CC1101 programmable radio.  

MQTT commands are used from Home Assistant as well as other controllers throughout my home.  It is also used to determine if sliding doors are open [not to lower the shades] by reading other MQTT sensors on the sliders.  Conversely if a slider is opened from the outside it will open the shade [I have 4 sliders each with a large single shade].

One issue i am having is OTA updates.  I’m not having much luck getting the ArduinoOTA library working w/HomeSpan.  Have you considered incorporating an OTA capability in the future? I’m planning on placing this unit in the attic - out of the way - OTA would be great for remote updates.

**Reply by RDunmow** [reply]
I had some issues with OTA at first, like MQTT I found that I can set it up, but only initialise it once the WiFi was connected, otherwise, the ESP would reboot.

**Reply by HomeSpan** [reply]
Thanks for the feedback.  I looked into OTA and its been on my list of things to incorporate into HomeSpan but I've not yet explored in detail.  One of my goals for HomeSpan is to allow end-users with no development experience to modify a device that has already been programmed without the need to plug the device into computer or perform any actual programming.  HomeSpan's WiFi Access point, which provides a web page to set WiFi Credentials and modify the HomeKit Setup Code, is an example.

For my Somfy shade controller, I'm thinking about a generic web page that would allow an end-user to "add" or "remove" channels to the bridge.  Normally this is done with a third-party App by the manufacturer of the controller, but doing this via a web page would make customization much more accessible for developers that don't want to program their own iPhone Apps (myself included).

The web pages themselves are the easy part, and it would be relatively straightforward to simply modify the WiFi Access Point pages in the HomeSpan library to add custom, device-specific set-up pages.  But this of course breaks the library for other uses.  So instead, I'm working through the best way to create hooks into the library and a web-page framework that would allow one to create custom set-up pages from within a sketch that the library automatically served when needed.  Should not be too difficult.

Re the Somfy shade controller, sensors on the shades themselves (or on a sliding door) is a robust solution.  As an second-best, I implemented some timing logic that tracks the cumulative time the shade rises versus lowers to estimate the position of the shade at any point.   I implemented this on a patio screen door and it works surprisingly well.  I can use the HomeKit Window Covering sliders and set the screen door to any point (30% open, 18% open, etc.).  An it's fun to tell Siri to open the door to 85% and watch it stop there automatically.  Of course this is not actual practical - it's a rolling screen on a patio, so not sure why I ever need to open it only 85% 😀.  But it is a good demonstration of how it would work with actual shades.

**Reply by HerbOtto** [reply]
I'm using parts of your somfy code for my vertical shades. Works great. I took a similar approach a while back with python-homekit which served me well.

I have dozens of ESP devices around hence the need for OTA. 

RDunmow: how did you get OTA to work? I got past the dreaded reboot cycle by delaying its setup much like MQTT but it's not discoverable. 
EDIT: I finally got OTA working - had to delay a few seconds after MQTT started - it's now discoverable! :-)

On another note: It would be great to be able to set the HomeSpan network name directly rather than just changing the prefix to a randomized series of characters.  I prefer to use names like "esp-somfy" throughout my network to find devices.

**Reply by RDunmow** [reply]
As I was having trouble with the wifi connectivity, I put all the OTA code in a small .h file and then would simply call a loop function in that code that would 

1) Check the WiFi was connected
2) check to see if the OTA code had been initialised (set a boolean flag when it had)
3) if wifi was enabled and the OTA code was initialised, I would then call the OTA handle function in the main loop.

This allowed the code to initialise later if the device hadn't been configured for WiFi and needed to go through that setup.

This was a quick piece of code that I need to revisit to make it a little more elegant, but it worked, now it's working I can fix later.

**Reply by HerbOtto** [reply]
That's exactly what I did as well - seems to work fine - reboots are consistent as well.

So in addition to HomeSpan I have MQTT, OTA and RemoteDebug [for additional command and status capability] installed.
